pesto/objective.py

```diff
@@ -1,30 +1,309 @@
+                return fval
+            elif sensi_orders == (1,):
+                if self.grad is True:
+                    grad = self.fun(x)[1]
+                else:
+                    grad = self.grad(x)
+                return grad
+            elif sensi_orders == (2,):
+                if self.hess is True:
+                    hess = self.fun(x)[2]
+                else:
+                    hess = self.hess(x)
+                return hess
+            elif sensi_orders == (0, 1):
+                if self.grad is True:
+                    fval, grad = self.fun(x)[0, 1]
+                else:
+                    fval = self.fun(x)
+                    grad = self.grad(x)
+                return fval, grad
+            elif sensi_orders == (1, 2):
+                if self.hess is True:
+                    grad, hess = self.fun(x)[1, 2]
+                else:
+                    hess = self.hess(x)
+                    if self.grad is True:
+                        grad = self.fun(x)[1]
+                    else:
+                        grad = self.grad(x)
+                return grad, hess
+            elif sensi_orders == (0, 1, 2):
+                if self.hess is True:
+                    fval, grad, hess = self.fun(x)[0, 1, 2]
+                else:
+                    hess = self.hess(x)
+                    if self.grad is True:
+                        fval, grad = self.fun(x)[0, 1]
+                    else:
+                        fval = self.fun(x)
+                        grad = self.grad(x)
+                return fval, grad, hess
+
+            else:
+                raise ValueError("These sensitivity orders are not supported.")
+        else:
+            raise ValueError("This mode is not supported.")
+
+    def get_fval(self, x):
+        fval = self.__call__(x, (0,), Objective.MODE_FUN)
+        return fval
+
+    def get_grad(self, x):
+        grad = self.__call__(x, (1,), Objective.MODE_FUN)
+        return grad
+
+    def get_hess(self, x):
+        hess = self.__call__(x, (2,), Objective.MODE_FUN)
+        return hess
+
+    def get_hessp(self, x, p):
+        hess = self.__call__(x, (2,), Objective.MODE_FUN)
+        return np.dot(hess, p)
+
+    def get_res(self, x):
+        res = self.__call__(x, (0,), Objective.MODE_RES)
+        return res
+
+    def get_sres(self, x):
+        sres = self.__call__(x, (1,), Objective.MODE_RES)
+        return sres
 
-    def call(par, sensi_order=0):
-        return nllh
 
 class AmiciObjective(Objective):
-    
-    def __init__(amici_model, amici_solver, edata, dim, sensi_order):
+    """
+    This is a convenience class to compute an objective function from an
+    AMICI model.
+
+    Parameters
+    ----------
+
+    amici_model: amici.model
+        The amici model.
+
+    amici_solver: amici.solver
+        The solver to use for the numeric integration of the model.
+
+    edata:
+        The experimental data.
+
+    max_sensi_order: int
```

Comment:
At least for second order: `model.o2mode` is one of `amici.AMICI_O2MODE_NONE`, `amici.AMICI_O2MODE_DIR`, `amici.AMICI_O2MODE_FULL`. 
Not sure if it is safe to assume we can always get first order sensitivities. But not having them would be rather the exception. So i would default to `max_sensi_order = 2 if model.o2mode else 1`

Reply:
please take into account that we can compute the `FIM` as second order approximation!

Reply:
but maybe we do not want to use it, e.g. because it's too unreliable, so in this case, the objective function should not return any value so that the optimization does not silently run through using wrong values.